src/differential_privacy/protect_grouping_keys.rs

```diff
@@ -67,51 +30,99 @@ impl Reduce {
     }
 
     // Returns a `Relation` outputing all grouping keys that can be safely released
-    pub fn grouping_values(&self, epsilon: f64, delta: f64) -> Result<Relation> {
+    pub fn grouping_values(
+        &self,
+        protected_entity_id: &str,
+        protected_entity_weight: &str,
+        epsilon: f64,
+        delta: f64,
+    ) -> Result<DPRelation> {
         let grouping_cols = self.grouping_columns()?;
         if !grouping_cols.is_empty() {
             PEPRelation::try_from(self.inputs()[0].clone().filter_fields(|f| {
-                grouping_cols.contains(&f.to_string()) || f == PE_ID || f == PE_WEIGHT
+                grouping_cols.contains(&f.to_string())
+                    || f == protected_entity_id
+                    || f == protected_entity_weight
             }))?
             .released_values(epsilon, delta)
         } else {
-            Err(Error::NoGroupingColumnsError)
+            Err(Error::GroupingKeysError("No grouping keys.".to_string()))
         }
     }
 
-    fn join_with_grouping_values(self, grouping_values: Relation) -> Result<Relation> {
-        let on: Vec<Expr> = grouping_values
-            .schema()
-            .iter()
-            .map(|f| {
-                Expr::eq(
-                    Expr::qcol(self.name().to_string(), f.name().to_string()),
-                    Expr::qcol(grouping_values.name().to_string(), f.name().to_string()),
+    pub fn protect_grouping_keys(
+        self,
+        protected_entity_id: &str,
+        protected_entity_weight: &str,
+        epsilon: f64,
+        delta: f64,
+    ) -> Result<(PEPRelation, PrivateQuery)> {
+        Ok(
+            if self.grouping_columns()? == vec![protected_entity_id.to_string()] {
```

Comment:
because the self is a PEPRelation so that we have added PE_ID in the grouping columns 
And by definition we cannot reveal the keys with only one PE_ID value